Source/WebCore/rendering/RenderElement.cpp

```diff
@@ -900,18 +900,15 @@ void RenderElement::styleWillChange(StyleDifference diff, const RenderStyle& new
         }
 
         // Keep layer hierarchy visibility bits up to date if visibility or skipped content state changes.
-        bool wasVisible = m_style.usedVisibility() == Visibility::Visible && !m_style.hasSkippedContent();
-        bool willBeVisible = newStyle.usedVisibility() == Visibility::Visible && !newStyle.hasSkippedContent();
-        if (wasVisible != willBeVisible) {
+        if ((m_style.usedVisibility() == Visibility::Visible) != (newStyle.usedVisibility() == Visibility::Visible)) {
+            if (CheckedPtr layer = enclosingLayer())
+                layer->dirtyVisibleContentStatus();
+        }
+
+        if (m_style.hasSkippedContent() != newStyle.hasSkippedContent()) {
             if (CheckedPtr layer = enclosingLayer()) {
-                if (willBeVisible) {
-                    auto* renderBox = dynamicDowncast<RenderBox>(*this);
-                    if (m_style.hasSkippedContent() && renderBox && isSkippedContentRoot(*renderBox))
-                        layer->dirtyVisibleContentStatus();
-                    else
-                        layer->setHasVisibleContent();
-                } else if (layer->hasVisibleContent() && (this == &layer->renderer() || layer->renderer().style().usedVisibility() != Visibility::Visible))
-                    layer->dirtyVisibleContentStatus();
+                for (RenderLayer* child = layer->firstChild(); child; child = child->nextSibling())
+                    child->dirtyVisibleContentStatus();
```

Comment:
any reason not marking the enclosing layer dirty? e.g. the entire layer goes hidden when content-visibility changes from 'visible' to 'hidden' (`m_style.hasSkippedContent() != newStyle.hasSkippedContent()` is true)

Reply:
Changed this a bit, once I understood the `isSkippedRootOrSkippedContent()` a bit better.

It now just marks the layer directly, if the used value of content-visibility directly (and since that's inherited for hidden subtrees, marks all the actually-hidden children layers).

In theory we don't need to mark the skipped-root as having dirty visible content, but it looks like we can't easily distinguish those from this caller.